RFC for Open Telemetry Implementation for Presto. (#33)
prestodb/rfcs: RFC-0011-open-telemetry.md

# **RFC0011 for Presto**

## Enhancing Open Telemetry Implementation in Presto
Proposers
* Suresh Babu Areekara
* Siddarth Ajay
* Ben Tony Joe

## [Related Issues]
* https://github.com/prestodb/presto/issues/23975

## Summary

> The existing Open Telemetry implementation https://github.com/prestodb/presto/pull/18534 was an experimental feature, had a limited set of telemetry data(Query state changes) and did not include a child span concept. The recent implementation will make Presto more flexible, allowing support for both parent and child spans. Additionally, traces can now be propagated to the worker nodes as well.

Reviewer: This should summarize what the document is proposing, there can be a sample implementation, but the design is more important than implementation in the RFC. We need to highlight the shortcomings of the current tracing provider SPI and propose changes to the current SPI to support the desired features.

It seems like the desired features are:

1. child spans
2. tracing propagated to workers

In the later sections we need to explain why the current interface does not allow these two features.

If there are more features required, they should be called out explicitly

```suggestion
Presto's original tracing implementation was merged in https://github.com/prestodb/presto/pull/18534. It was an experimental feature, had a limited set of telemetry data, and did not include a child span concept. This document proposes an expansion to the tracing provider SPI to

1. Support child spans in traces to make understanding nested operations simpler
2. Allow traces to be propagated to workers so that the entire lifespan of a query can be traced from analysis through execution
```

Author: Added.

## Background

> OpenTelemetry is a powerful serviceability framework that helps to gain insights into the performance and behaviour of the systems. It facilitates generation, collection, and management of telemetry data such as traces.

Reviewer: We should drop the term "Open Telemetry" as much as possible from the document because we're trying to implement an SPI interface that open telemetry is capable of implementing. The SPI tracing provider interface is the important thing to highlight.

> ### Existing Open Telemetry

Reviewer: We should update this to describe the current `TracerProvider` interface and specifically describe why it is insufficient

Author: Updated.

![Traces existing implementation](/RFC-0011-open-telemetry/traces-existing-implementation-oss-presto.png)

> The OSS Presto had a basic implementation of Open Telemetry with few limitations listed below.

Reviewer: ```suggestion
Presto has a basic `TracerProvider` interface which makes it difficult to implement the following desirable features:
```

Author: Updated.

- Have only less number of spans and covers only query state transitions like dispatching, planning, analyzing, and
scheduling. ie. it does not cover most of the internal operations during a query execution.
- No additional information is being passed as query attributes or events.
- No correlation between traceId and queryId.
- Does not identify failed queries
- No context propagation to track worker nodes.

> 

Reviewer: Here you should point out specifically which SPI methods used for tracing make implementing the particular features difficult

## Proposed Implementation
With additional changes to the OpenTelemetry SPI, we get the following advantages -
- Presto can be manually instrumented which gives more flexibility and control. Easier to customize what operations can be monitored.
- More number of spans with parent and child span concept which helps to get the insight on internal operations.
- Able to pass queryId as attribute so that traces and query get connected each other.
- Can identify failed queries and filter it out.
- Ability to pass additional information as span attributes and events.
- Context propagation from coordinator to worker nodes.

> 

Reviewer: This section needs to spell out exactly what changes are being made to the `TracerProvider` and `Tracer` interfaces in java code, and also any additional architectural changes, such as things that the C++ and Java workers might have to implement. The diagrams are good, but we need to understand the exact changes to the existing interfaces. It isn't clear what the changes are just looking at the diagrams.

I see there is an SPI section below, but it needs to be moved up here and the exact changes to the flow of tracing spans/blocks must be described in text. The descriptions also need to spell out why the existing `Tracer`/`TracerProvider` interfaces are insufficient and how the new SPI changes make the desired features possible

Author: Hi @ZacBlanco Moved the SPI changes as per the comment https://github.com/prestodb/rfcs/pull/33#discussion_r1977179341 . Please let me know where to keep. Under Adoption Plan or Proposed implementation ?

Author: Updated what all additions in instrumentation flow with the new proposal. Context propagation from Coordinator to Worker is entirely new for the proposed implementation.

- Propagation is usually handled by instrumentation libraries and is transparent to the user. In the event that you need to manually propagate context, you can use the Propagators API.

- OpenTelemetry maintains several official propagators. The default propagator is using the headers specified by the W3C TraceContext specification.
  - In Presto in areas where REST calls involved, we use the header for context propagation as per the above image.
  - Presto Coordinator fetch the current span context and inject as the traceparent http header. Which is then extracted from the Worker side and use to create the child spans with the parent context.
  - In some other areas parent context is available in child context and we directly set the parent context in child spans.

### Low level design
![Trace flow](/RFC-0011-open-telemetry/open-telemetry-trace-flow-diagram.png)

## [Optional] Other Approaches Considered
Based on the discussion, this may need to be updated with feedback from reviewers.

## Adoption Plan

> ### SPI Changes

Reviewer: The SPI changes need to be moved into the `Proposed Implementation` section

Author: Hi @ZacBlanco @aaneja Moved SPI changes as per the comment https://github.com/prestodb/rfcs/pull/33#discussion_r1977179341 . Please let me know where to keep. Under Adoption Plan or Proposed implementation ?

```java
class BaseSpan {}
```
SPI can be extended with the specific implementation which contains the actual SDK Span. BaseSpan propagates through the main presto code. Introduced as part of new design.

> #### Additions

Reviewer: Please consolidate the `Additions` section into a single one

Author: Done.

```java
default void close()
{
    return;
}

default void end()
{
    return;
}
```

```java
class Tracer {}
```
SPI for the span operations which can be implemented by specific serviceability framework.

#### Additions
```java

> void loadConfiguredOpenTelemetry();

Reviewer: The SPI cannot have OpenTelemetry-specific methods.

Author: Hi @ZacBlanco @aaneja Moved SPI changes as per the comment https://github.com/prestodb/rfcs/pull/33#discussion_r1977179341 . Please let me know where to keep. Under Adoption Plan or Proposed implementation ?

Author: Updated.

Runnable getCurrentContextWrap(Runnable runnable);
boolean isRecording();
Map<String, String> getHeadersMap(T span);
void endSpanOnError(T querySpan, Throwable throwable);
void addEvent(T span, String eventName);
void setAttributes(T span, Map<String, String> attributes);
void recordException(T span, String message, RuntimeException runtimeException, ErrorCode errorCode);
void setSuccess(T span);
T getInvalidSpan();
T getRootSpan(String traceId);
T getSpan(String spanName);
U scopedSpan(String name, Boolean... skipSpan);

> Optional<String> spanString(T span);

Reviewer: These all should have javadoc comments explaining what they do. Ideally the RFC should also explain how and where they are used.

Author: Updated.

```

#### Deletions
```java
String tracerName = "com.facebook.presto";
void addPoint(String annotation);
void startBlock(String blockName, String annotation);
void addPointToBlock(String blockName, String annotation);
void endBlock(String blockName, String annotation);
void endTrace(String annotation);

> ```

Reviewer: We cannot just delete from the SPI, so we must add to the transition plan how we will either (1) maintain compatibility with the existing tracing methods and (2) eventually transition to the new API entirely. Will there be a feature flag?

Author: @ZacBlanco @aaneja As the existing OTel feature was and experimental one with limited spans, coexistence of both is relevant and a feasible approach?

```java
class TraceProvider {}
```

> SPI to supply different Tracer implementations.

Reviewer: We cannot delete this because we *must* be able to supply multiple tracer implementations

Author: TraceProvider completely not removed. only some methods in it.

#### Additions
```java
T create();
```

#### Deletions
```java
String getTracerType();
Function<Map<String, String>, TracerHandle> getHandleGenerator();
Tracer getNewTracer(TracerHandle handle);
```

```java
class NoopTracer {}
```

> Removed as we have a default Tracer implementation in presto-main/..tracing/DefaultTelemetryTracer

Reviewer: The tracer implementation should not have any components in `presto-main`

#### Deletions
```java
public void addPoint(String annotation)
{
}

@Override
public void startBlock(String blockName, String annotation)
{
}

@Override
public void addPointToBlock(String blockName, String annotation)
{
}

@Override
public void endBlock(String blockName, String annotation)
{
}

@Override
public void endTrace(String annotation)
{
}

@Override
public String getTracerId()
{
    return "noop_dummy_id";
}
```

```java
class TracerHandle {}
```
Removed as not required for the current implementation.

#### Deletions
```java
String getTraceToken();
```

### Configuration

> Trace can be configured by modifying the values in presto-main/etc/telemetry-tracing.properties

Reviewer: This seems specific to the OpenTelemetry implementation. I think it's fine to propose changes to how we configure the existing OpenTelemtry tracing plugin, but we need to be very clear about the any configuration properties that would be added to `TracingConfig` since it affects `config.properties`

Author: Thanks for the feedback! We'll make sure to clearly document it. Also Added the current configuration along with proposed configuration in RFC.

2. Allow traces to be propagated to workers so that the entire lifespan of a query can be traced from analysis through execution

## Background
TracerProvider/Tracer can be extended and implemented with serviceability frameworks. It facilitates generation, collection, and management of telemetry data such as traces.

### Existing implementation
![Traces existing implementation](/RFC-0011-open-telemetry/traces-existing-implementation-oss-presto.png)

Presto has a basic `TracerProvider` and `Tracer` interfaces which makes it difficult to implement the following desirable features:
- Have only limited number of spans and covers only query state transitions like dispatching, planning, analyzing, and
scheduling. ie. it does not cover most of the internal operations during a query execution(The implementation for methods startBlock and endBlock are based on Map and cannot be used to handle large number of spans as the span name can be same for many.).
- No additional information is being passed as query attributes or events. No suitable methods in SPI to set events and attributes.
- No correlation between traceId and queryId. No suitable methods in SPI to set queryId.

> - Does not identify failed queries. No suitable methods in SPI to pass Error/Exception information and deals with failed query spans.

Reviewer: ```suggestion
- Does not identify failed queries. No suitable methods in SPI to pass Error/Exception information and details with failed query spans.
```

Is this supposed to be "details"?

- OpenTelemetry maintains several official propagators. The default propagator is using the headers specified by the W3C TraceContext specification.
  - In Presto in areas where REST calls involved, we use the header for context propagation as per the above image.
  - Presto Coordinator fetch the current span context and inject as the traceparent http header. Which is then extracted from the Worker side and use to create the child spans with the parent context.
  - In some other areas parent context is available in child context and we directly set the parent context in child spans.

### Low level design
![Trace flow](/RFC-0011-open-telemetry/open-telemetry-trace-flow-diagram.png)

## [Optional] Other Approaches Considered
Based on the discussion, this may need to be updated with feedback from reviewers.

## Adoption Plan
### SPI Changes
#### Additions
```java
class BaseSpan {}
```
SPI can be extended with the specific implementation which contains the actual SDK Span. BaseSpan propagates through the main presto code. Introduced as part of new design.

```java

>   default void close()

Reviewer: are these methods on the class of `BaseSpan`? It is not clear. Also, `default` is not valid unless `BaseSpan` is defined as an `interface`.

How do the methods `close()` and `end()` differ? When should they be used?

Author: BaseSpan is an SPI interface which is autoclosable. close() came from `AutoCloseable` and end() part of BaseSpan.
close() can be used to end span in a try-with-resource and end() to manually end the span.

  {
      return;
  }

    default void end()
  {
      return;
  }
```

```java
class Tracer {}
```

> SPI for the span operations which can be implemented by specific serviceability framework.

Reviewer: If this is meant to be a comment on the `Tracer` class, can you add it as a javadoc above the class within the code block?

Author: Updated.

```java
  /**
   * Instantiates required Telemetry instances after loading the plugin implementation.
   */

>   void loadConfiguredTelemetry();

Reviewer: Is this necessary as an interface method on the SPI? Why can't initialization just occur in the constructor when the plugin is loaded?

Author: @ZacBlanco Thanks for the feedback. That's a valid point. Removed method from SPI.

  /**
   * Gets current context wrap.
   *
   * @param runnable the runnable
   * @return the current context wrap
   */

>   Runnable getCurrentContextWrap(Runnable runnable);

Reviewer: What is a context wrap?

Author: It is Context.current().wrap(runnable); which is used to propagate the current context to a new thread or async task.
It ensures that the trace context is correctly propagated when executing the Runnable in a different thread.

  /**
   * Returns true if this Span records tracing events.
   *
   * @return the boolean
   */

>   boolean isRecording();

Reviewer: Under what circumstances should `isRecording` return false?

  /**
   * Returns headers map from the input span.
   *
   * @param span the span
   * @return the headers map
   */
  Map<String, String> getHeadersMap(T span);

  /**
   * Ends span by updating the status to error and record input exception.
   *
   * @param querySpan the query span
   * @param throwable the throwable
   */
  void endSpanOnError(T querySpan, Throwable throwable);

  /**
   * Add the input event to the input span.
   *
   * @param span      the span
   * @param eventName the event name
   */
  void addEvent(T span, String eventName);

  /**
   * Sets the attributes map to the input span.
   *
   * @param span       the span
   * @param attributes the attributes
   */
  void setAttributes(T span, Map<String, String> attributes);

  /**
   * Records exception to the input span with error code and message.
   *
   * @param span        the query span
   * @param message          the message
   * @param runtimeException the runtime exception
   * @param errorCode        the error code
   */

>   void recordException(T span, String message, RuntimeException runtimeException, ErrorCode errorCode);

Reviewer: Can we use `Exception` instead of `RuntimeException` to make this more generic?

Also, this method seems similar to `addEvent`. It seems more like a convenience method. We could define the interface in a different way that would make it more generic and allow flexibility in the event implementation. If we remove `recordException` but maintain `addEvent` we could have

```java

public void addEvent(T span, Event event);

public interface Event {}

public class ExceptionEvent implements Event {
    private final String message;
    private final Exception exception;
    ....
}

public class GenericEvent implements event {
    private final String eventName;
    ....
}

```

Author: Hi @ZacBlanco Thanks for the feedback. Updated accordingly.

  /**
   * Sets the status of the input span to success.
   *
   * @param span the query span
   */
  void setSuccess(T span);

  /**
   * Returns an invalid Span. An invalid Span is used when tracing is disabled.
   *
   * @return the invalid span
   */

>   T getInvalidSpan();

Reviewer: Why do we need a different method for this? If tracing is disabled, the implementation should handle returning an implementation of T which is an invalid span. Otherwise, users would need to write a conditional every time this is called?

Author: I think yes the method is needed @ZacBlanco 
If plugin loaded  & tracing disabled  --> no-op sdk span
If plugin not loaded  & tracing enabled/disabled  --> default implementation of <T>.

  /**
   * Creates and returns the root span.
   *
   * @return the root span
   */

>   T getRootSpan(String traceId);

Reviewer: Does it always create it? Or only if it doesn't exist yet?

Author: For each query 1 root span will always be created.

  /**
   * Creates and returns the span with input name.
   *
   * @param spanName the span name
   * @return the span
   */
  T getSpan(String spanName);

  /**
   * Creates and returns the ScopedSpan with input name.
   *
   * @param name     the name

>    * @param skipSpan the skip span

Reviewer: what is a `skipSpan?

Author: skipSpan corresponds to `span-sampling` in telemetry-tracing.properties to skip the spans which are less required. These were added to reduce the number of spans so that we can improve the performance and increase throughput, especially in high-load environments.

   * @return the u
   */
  U scopedSpan(String name, Boolean... skipSpan);

  /**
   * Returns the span info as string.
   *
   * @param span the span
   * @return the optional
   */

>   Optional<String> spanString(T span);

Reviewer: Why is there a separate method for this? Why can't this just be the implementation of `T.toString()`?

Author: Thanks for the feedback. Moved the implementation.

```

```java
class TraceProvider {}
```

> SPI to supply different Tracer implementations.

Reviewer: Similar comment on moving this to a javadoc.

Author: Updated.

```java
  T create();

> ```

Reviewer: If this belongs on `TracerProvider` please put it in the same code block

Author: Updated.

#### Deletions
```java
class Tracer {}
```

```java
  String tracerName = "com.facebook.presto";
  void addPoint(String annotation);
  void startBlock(String blockName, String annotation);
  void addPointToBlock(String blockName, String annotation);
  void endBlock(String blockName, String annotation);
  void endTrace(String annotation);

> ```

Reviewer: are we deleting the entire class or just some methods? If it's the whole class you can just list that we are deleting the `Tracer` class

Author: It's we are deleting only some methods. Updated acordingly.
